core/src/main/java/com/yahoo/oak/HashChunk.java

```diff
@@ -0,0 +1,413 @@
+     */
+    HashChunk(int maxItems, AtomicInteger externalSize, MemoryManager vMM, MemoryManager kMM,
+        OakComparator<K> comparator, OakSerializer<K> keySerializer,
+        OakSerializer<V> valueSerializer, UnionCodec hashIndexCodec) {
+
+        super(maxItems, externalSize, comparator);
+        assert Math.pow( 2, hashIndexCodec.getFirstBitSize() ) <= maxItems ;
+        assert hashIndexCodec != null;
+
+        this.hashIndexCodec = hashIndexCodec;
+        this.entryHashSet = // must be called after setSecondLevelBitsThreshold
+            new EntryHashSet<>(vMM, kMM, getMaxItems(), keySerializer, valueSerializer,
+                comparator);
+    }
+
+    /**
+     * Create a child HashChunk where this HashChunk object as its creator.
+     */
+    HashChunk<K, V> createChild(UnionCodec hashIndexCodec) {
+        HashChunk<K, V> child =
+            new HashChunk<>(getMaxItems(), externalSize,
+                entryHashSet.valuesMemoryManager, entryHashSet.keysMemoryManager,
+                comparator, entryHashSet.keySerializer, entryHashSet.valueSerializer,
+                hashIndexCodec);
+        updateBasicChild(child);
+        return child;
+    }
+
+    private int calculateKeyHash(K key) {
+        int hashKey = key.hashCode(); // hash can be positive, zero or negative
+        return Math.abs(hashKey); // EntryHashSet doesn't accept negative hashes
+    }
+    /********************************************************************************************/
+    /*-----------------------------  Wrappers for EntryOrderedSet methods -----------------------------*/
+
+    /**
+     * See {@code EntryOrderedSet.isValueRefValidAndNotDeleted(int)} for more information
+     */
+    boolean isValueRefValidAndNotDeleted(int ei) {
+        return entryHashSet.isValueRefValidAndNotDeleted(ei);
+    }
+
+    /**
+     * See {@code EntryOrderedSet.readKey(ThreadContext)} for more information
+     */
+    void readKey(ThreadContext ctx) {
+        entryHashSet.readKey(ctx);
+    }
+
+    /**
+     * See {@code EntryOrderedSet.readValue(ThreadContext)} for more information
+     */
+    void readValue(ThreadContext ctx) {
+        entryHashSet.readValue(ctx);
+    }
+
+    /**
+     * See {@code EntryOrderedSet.readKey(KeyBuffer)} for more information
+     */
+    boolean readKeyFromEntryIndex(KeyBuffer key, int ei) {
+        return entryHashSet.readKey(key, ei);
+    }
+
+    /**
+     * See {@code EntryOrderedSet.readValue(ValueBuffer)} for more information
+     */
+    boolean readValueFromEntryIndex(ValueBuffer value, int ei) {
+        return entryHashSet.readValue(value, ei);
+    }
+
+    /**
+     * Writes the key off-heap and allocates an entry with the reference pointing to the given key
+     * See {@code EntryOrderedSet.allocateEntryAndWriteKey(ThreadContext)} for more information
+     */
+    boolean allocateEntryAndWriteKey(ThreadContext ctx, K key) {
+        int keyHash = calculateKeyHash(key);
```

Comment:
To avoid recalculating the hash for recurrent attempts, can we save it in the `ThreadContext`?

Reply:
It is to be done in via InternalOakHashMap, therefore yet separated between FirstLevelHashArray and HashChunk, but I added the start of this work.